### DISCUSSION TITLE
Sparql code for dps

### DISCUSSION THREAD

**Discussion by TedofSwe** [OP-TED/ted-sparql-editor#9] [Q&A] [answered]
Hello,

Many thanks for the initiative with Ted Open Data Service!

can anyone please help me to create a SparQL-query so I can see all Dynamic purchasing systems established in EU?
I have tried myself but I'm not a programmer so I've failed.

I don't know how big of a ask this is. But if it's a big thing I'm grateful just to be pointed in the right direction. 
I'm looking at the model glossary for the The eProcurement Ontology and there I find "epo;hasDPSScope". But maybe I'm looking at the wrong "endpoint".

Thanks!
Ted/Sweden

**Comment by cristianvasquez** [comment, marked as answer]
Dear @TedofSwe , 

Currently, there is no code list available that enumerates the different DPS used in the EU. As a result, it is not possible to perform a SPARQL query to retrieve them.

The property `epo:hasDPScope` is designed to describe [DPS usage](https://op.europa.eu/en/web/eu-vocabularies/concept-scheme/-/resource?uri=http://publications.europa.eu/resource/authority/dps-usage), but this may not align with the specific information you are looking for.

![](http://www.plantuml.com/plantuml/img/ZP5BQyCm48Jl-XMBxWhfrOGGsXnx2BWvX132LbviDteuMgabfFtljKoXURJWFOePpRUpOi4HGqpM23wqhiE05lph6-a8Pm7zFA8-rC4dLprvumCyjtKJDsIC_n1VGXja1acTBvOdXxRLgnHqWqpbYIFP5zADQu-9remrIK26aTeXFcjxmN9Lbjft10LWLEzU5rN7AWsoxNGoUTZrA9Si_yJU9MGRkMH1tluxU33kVnxtwTVnPyBiKwdnFOZnZIar1vbtb1docaVsXTnK91RagdpERm00)

I hope this helps clarify your question.

Best regards,
Cristian

**Reply by TedofSwe** [reply]
Dear Cristian,
Thank you very much for your response. Much appreciated.
Best regards, Ted

**Comment by VladimirAlexiev** [comment]
HI @tedofswe and @cristianvasquez !
I'd guess there is a specific procedure type to indicate DPS but I don't know it.
However, we can explore from the given property to find some DPSes.

You can try the queries below at https://yasgui.org/ by specifying https://publications.europa.eu/webapi/rdf/sparql as the SPARQL endpoint.
(Since TED's own SPARQL Editor is lacking)

- Let's first count things with `hasDPSScope`. These are **DynamicPurchaseSystemTechnicalUsage**:
```sparql
SELECT (count(*) as ?c) WHERE {
  ?x epo:hasDPSScope ?y
} # 4.1M
```
- Now let's count things that refer to such "DynamicPurchaseSystemTechnicalUsage". These are **Lots**. If you use `distinct` there's again 4.1M but a few less than the count without distinct.
```sparql
SELECT (count(distinct ?x) as ?c) WHERE {
  ?x epo:usesTechnique/epo:hasDPSScope ?y
} # 4.1M
```
- Now let's find Notices comprised of these Lots. There are 2 relations for that, so let's first count them:
```sparql
PREFIX epo: <http://data.europa.eu/a4g/ontology#>
select (count(?y) as ?refersToLot) (count(?z) as ?announcesLot) (count(distinct ?y) as ?refersToLotDistinct) (count(distinct ?z) as ?announcesLotDistinct) {
  {?x epo:refersToLot ?y}
  union {?x epo:announcesLot ?z}
}
```
|   | refersToLot             | announcesLot            | refersToLotDistinct     | announcesLotDistinct    |
|---|-------------------------|-------------------------|-------------------------|-------------------------|
| 1 | "4617897" ^^xsd:integer | "2250689" ^^xsd:integer | "4582459" ^^xsd:integer | "2227314" ^^xsd:integer |

So I assume that `refers` is a broader (weaker) relation and `announces` is a narrower (stronger) relation subsumed by `refers`. I will therefore use `refers`:
- Now let's find the Notices that refer to Lots that use technique that has DPS scope (quite a mouthful)!
```sparql
PREFIX epo: <http://data.europa.eu/a4g/ontology#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT (count(distinct ?x) as ?distinct) (count(?x) as ?withDups) WHERE {
  ?x epo:refersToLot/epo:usesTechnique/epo:hasDPSScope ?y
} 
```
The `distinct` is quite necessary here since the same DPS Notice may have multiple lots (about 3.9x on average):
|   | distinct                | withDups                |
|---|-------------------------|-------------------------|
| 1 | "1250980" ^^xsd:integer | "4426892" ^^xsd:integer |

Now @tedofswe, what do you want to do with these 1.25M DPSes?
The next step is to find out which ones are still active...

**Reply by TedofSwe** [reply]
Thank you so much! The next step is to list the values in the Notices and extract specific fields of interest for analysis. However, I already downloaded all Notices this summer and wrote Python scripts to do this. I know very little about SPARQL, so I’m not sure whether this approach is better than the one I’ve built. I had to create separate scripts for the TED XML schema and for the new eForms (from October 25th 2023).

Thanks again!

**Reply by VladimirAlexiev** [reply]
The original mission of TED ODS is to make access easier because all the data is integrated and easier to query. 

From your evidence and the low interest/number of discussions here, it seems OP is still a long way from achieving that mission...
